Add "Leaky Bucket Rate Limiting" Blogpost (#45)
nilenso/blog.nilenso.com: source/_posts/2024-01-13-leaky-bucket.markdown

---
title: "Leaky Bucket Rate Limiting"
kind: article
created_at: 2024-01-13 19:40:00 UTC
author: Shivam Singhal
layout: post
---
Hmm…leaky bucket? The first time I heard these words, it made me chuckle. I thought what kind of name is this for a rate limiting strategy.

> But upon exploring it, it’s rather a very simple, elegant, and popular strategy used in a lot of production environments.

Reviewer: Prefer active voice, since the rest of the article is like that.
```suggestion
But upon exploring it, I found it to be a rather simple and elegant strategy used in a lot of production environments.
```

> But what are we rate limiting? It’s obvious that we want to rate limit the incoming requests. But do we rate limit the whole server(due to hardware constraints) or do we rate limit per user?
> 
> The answer, well, depends on your business requirement. How do you differentiate between different users? An easy way out for us is that we can rate limit based on the IP Address of the request and that’s what we are going to do.

Reviewer: There's a jump betweem "what are we rate limiting" to "it's obvious we're limiting requests". Here's an attempt to explain we're limiting requests a bit.

```suggestion
But what are we rate limiting? The answer, well, depends on your business requirement. A very common use case is to rate limit requests from a user of your service in a given period of time.

How do you differentiate between different users? An easy way out for us is that we can rate limit based on the IP Address of the request and that’s what we are going to do.
```

<aside>

> 💡 Bonus: check out an implementation of leaky bucket at the bottom!

Reviewer: This sounds like it's trying to drive engagement a bit like how `watch till the end of this video` does 😅. It's alright, but try saying it in a way where you can avoid that connotation?

Author: True. Let me try to remove this connotation.

### Here's how it works:

1. **Bucket as a Buffer**: Imagine a bucket that represents a buffer for incoming requests. The size of this bucket determines the buffer capacity, which means how many requests it can hold at any given time.
2. **Constant Requests Outflow**: Requests leave the bucket (or buffer) at a constant rate, regardless of the rate at which they arrive. This constant outflow rate ensures that data is processed at a steady, manageable pace per user.
3. **Incoming Requests**: Requests arrive at the bucket at varying rates. If the incoming rate is less than or equal to the outflow rate, packets are processed smoothly.
4. **Overflow Management**: When the incoming rate exceeds the outflow rate, excess packets start filling the bucket. If the bucket fills up, new incoming packets are discarded and return a `429 - TOO MANY REQUESTS`  until there is space in the bucket again for the IP address.

### **Advantages**

The Leaky Bucket algorithm is effective in smoothing out bursty traffic - turning a variable-rate input into a constant-rate output. It's also useful in ensuring that the data processing doesn't get overwhelmed during high traffic periods.

> ### **Applications**

Reviewer: The headings don't add much, the paragraphs stand on their own as well. Could make this a single paragraph.

It's widely used in packet-switched networks for congestion control, in APIs for rate limiting requests, and in various other contexts where it's essential to regulate the flow of data.

## Helpful article(s)

[https://redis.com/glossary/rate-limiting](https://redis.com/glossary/rate-limiting)

> The above article is highly recommended because it mentions about some other rate limited strategies like fixed window, sliding window and token bucket rate limiting

Reviewer: Suggested rephrasing from passive voice to active voice:

`I would highly recommend reading this redis article. It mentions ...`

Reviewer: This doesn't need to be a separate heading either, see if it fits as a sentence somewhere?
